osu.Game/Rulesets/Objects/HitObject.cs

```diff
@@ -104,10 +123,12 @@ public HitObject()
         /// <param name="cancellationToken">The cancellation token.</param>
         public void ApplyDefaults(ControlPointInfo controlPointInfo, BeatmapDifficulty difficulty, CancellationToken cancellationToken = default)
         {
+            DifficultyControlPoint = LocalDifficultyControlPoint ?? controlPointInfo.DifficultyPointAt(StartTime);
+
             ApplyDefaultsToSelf(controlPointInfo, difficulty);
 
             // This is done here since ApplyDefaultsToSelf may be used to determine the end time
-            SampleControlPoint = controlPointInfo.SamplePointAt(this.GetEndTime() + control_point_leniency);
+            SampleControlPoint = LocalSampleControlPoint ?? controlPointInfo.SamplePointAt(this.GetEndTime() + control_point_leniency);
```

Comment:
@smoogipoo I tried replacing this with your recommendation:

```csharp
        public void ApplyDefaults(ControlPointInfo controlPointInfo, BeatmapDifficulty difficulty, CancellationToken cancellationToken = default)
        {
            updateDifficultyControlpoint(controlPointInfo);
            updateSampleControlPoint(controlPointInfo);
        }

        private void updateSampleControlPoint(ControlPointInfo controlPointInfo)
        {
            if (SampleControlPoint == beatmapSampleControlPoint)
                SampleControlPoint = null;

            beatmapSampleControlPoint = controlPointInfo.SamplePointAt(this.GetEndTime() + control_point_leniency);

            SampleControlPoint ??= beatmapSampleControlPoint;
        }

        private void updateDifficultyControlpoint(ControlPointInfo controlPointInfo)
        {
            if (DifficultyControlPoint == beatmapDifficultyControlPoint)
                DifficultyControlPoint = null;

            beatmapDifficultyControlPoint = controlPointInfo.DifficultyPointAt(StartTime);

            DifficultyControlPoint ??= beatmapDifficultyControlPoint;
        }
```

But this feels a bit unsafe / hard to work with. The method I was using to decide whether a custom override control point is required (whether the `Local` version was `null` or not) can no longer be used. So instead, an alternative approach would be needed:
- Always make a new one on each change (probably not great?)
- Expose `bool`s in `HitObject` to allow access to the `DifficultyControlPoint == beatmapDifficultyControlPoint` comparison.

For reference, see the use case at https://github.com/ppy/osu/pull/10417/files#diff-7d2412aa7d34bb3e5c9b8ab573bf0bf6dd0795224a6261c46bd4a862eed68933R207-R219

Reply:
Creating a new point every change does work, so that is a direction we can take, potentially:

```csharp
                                var closestGlobalDifficulty = editorBeatmap.ControlPointInfo.DifficultyPointAt(hitObject.StartTime);

                                var localDifficulty = hitObject.DifficultyControlPoint = new DifficultyControlPoint
                                {
                                    SpeedMultiplier = hitObject.DifficultyControlPoint.SpeedMultiplier
                                };

                                localDifficulty.SpeedMultiplier *= (repeatHitObject.Duration / proposedDuration);

                                if (closestGlobalDifficulty.IsRedundant(localDifficulty))
                                    hitObject.DifficultyControlPoint = null;

                                beatmap.Update(hitObject);
```

But note that we are doing a second lookup on `ControlPointInfo` to check whether the local point is redundant. Also it's hard to know whether we should be displaying a local override any more, so every hitobject shows one as a result:

![20210817 184548 (dotnet)](https://user-images.githubusercontent.com/191335/129703621-80063ee7-aca2-43bf-b556-51480ce2ca23.png)

Previously this was also attained from the `null` case.

Reply:
I've pushed the changes above to https://github.com/ppy/osu/compare/master...peppy:smgi-control-point-suggestions?expand=1 if you want to have a look around.

Reply:
As discussed earlier, it may be better to go the full-hog and copy the control points anyway, removing the concept of "override" and fixing the timeline display to reference the previous hitobject as an additional step.

It's going to be difficult to encode the beatmap while maintaining the concept of an "override" anyway.